Mathlib/NumberTheory/Padics/WithVal.lean

```diff
@@ -0,0 +1,181 @@
+/-
+Copyright (c) 2025 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Analysis.RCLike.Basic
+import Mathlib.GroupTheory.ArchimedeanDensely
+import Mathlib.NumberTheory.Padics.PadicNumbers
+import Mathlib.Topology.Algebra.UniformRing
+import Mathlib.Topology.Algebra.Valued.ValuedField
+import Mathlib.Topology.Algebra.Valued.WithVal
+import Mathlib.Topology.GDelta.MetrizableSpace
+
+/-!
+# Equivalence between `ℚ_[p]` and `WithVal v`
+
+-/
+
+variable {p : ℕ} [Fact p.Prime]
+
+open NNReal WithZero UniformSpace
+
+-- TODO: use Rat.padicValuation after #27667
+lemma Padic.continuous_cast_withVal : Continuous ((Rat.castHom ℚ_[p]).comp
+    (WithVal.equiv ((Padic.mulValuation (p := p)).comap (Rat.castHom ℚ_[p]))).toRingHom) := by
+  rw [continuous_iff_continuousAt]
+  intro x U
+  simp only [RingEquiv.toRingHom_eq_coe, RingHom.coe_comp, Rat.coe_castHom, RingHom.coe_coe,
+    Function.comp_apply, Metric.mem_nhds_iff, gt_iff_lt, Filter.mem_map, Valued.mem_nhds,←
+    Set.image_subset_iff, forall_exists_index, and_imp]
+  intro ε hε hU
+  obtain ⟨y, hy⟩ := rat_dense p 0 hε
+  lift ε to ℝ≥0 using hε.le
+  -- inventing a non-zero element of the ball
+  have hp : (p : ℝ≥0)⁻¹ < 1 := by simp [inv_lt_one_iff₀, Nat.Prime.one_lt Fact.out]
+  obtain ⟨n, hn⟩ := exists_pow_lt₀ hp (Units.mk0 ε (by simpa using hε.ne'))
+  simp only [inv_pow, Units.val_mk0, ← NNReal.coe_lt_coe] at hn
+  use Units.mk0 (exp (-n : ℤ)) (by simp)
+  refine subset_trans ?_ hU
+  intro y
+  simp only [Units.val_mk0, Set.mem_image, Set.mem_setOf_eq,
+    Metric.mem_ball, dist_eq_norm, forall_exists_index, and_imp]
+  rintro y h rfl
+  set x' : ℚ := (WithVal.equiv
+    ((Padic.mulValuation (p := p)).comap (Rat.castHom ℚ_[p]))) x with hx
+  set y' : ℚ := (WithVal.equiv
+    ((Padic.mulValuation (p := p)).comap (Rat.castHom ℚ_[p]))) y with hy
+  change Valuation.comap _ _ (y' - x') < exp _ at h
+  refine hn.trans' ?_
+  simp only [← Rat.cast_sub, padicNormE.eq_padicNorm, padicNorm, zpow_neg, NNReal.coe_inv,
+    NNReal.coe_pow, NNReal.coe_natCast]
+  split_ifs with hxy
+  · simp only [Rat.cast_zero, inv_pos]
+    exact pow_pos (by simp [Nat.Prime.pos Fact.out]) _
+  · simp only [Rat.cast_inv, Rat.cast_zpow, Rat.cast_natCast]
+    rw [inv_lt_inv₀, ← zpow_natCast, zpow_lt_zpow_iff_right₀]
+    · simpa [← Rat.cast_sub, hxy] using h
+    · exact_mod_cast Nat.Prime.one_lt Fact.out
+    · exact zpow_pos (by simp [Nat.Prime.pos Fact.out]) _
+    · exact pow_pos (by simp [Nat.Prime.pos Fact.out]) _
+
+lemma Padic.isUniformInducing_cast_withVal : IsUniformInducing ((Rat.castHom ℚ_[p]).comp
+    (WithVal.equiv ((Padic.mulValuation (p := p)).comap (Rat.castHom ℚ_[p]))).toRingHom) := by
+  constructor
```

Comment:
This proof uses (twice) a constructor to produce a term of a one-field structure, which is almost a defeq abuse. Is there a way you can use `Filter.HasBasis.isUniformInducing_iff`(together with `basis_sets`) to avoid this?